File: golang/2434.using-a-robot-to-print-the-lexicographically-smallest-string.go
func robotWithString(s string) string {
	n := len(s)
	cnt := make([]int, 26)
	res := make([]byte, 0, n)
	st := make([]byte, 0, n/4)

	for _, c := range s {
		cnt[c-'a']++
	}

	i := 0

	for c := byte('a'); c <= 'z'; c++ {
		// pop from stack all the characters that are less than or equal to current character
		for len(st) > 0 && st[len(st)-1] <= c {
			res = append(res, st[len(st)-1])
			st = st[:len(st)-1]
		}

		for cnt[c-'a'] > 0 {
			if s[i] == c {
				res = append(res, s[i])
			} else {
				st = append(st, s[i])
			}

			cnt[s[i]-'a']--
			i++
		}
	}

	for len(st) > 0 {
		res = append(res, st[len(st)-1])
		st = st[:len(st)-1]
	}

	return string(res)
}

File: python/2434.using-a-robot-to-print-the-lexicographically-smallest-string.py


class Solution:
    def robotWithString(self, s: str) -> str:
        stack = []

        count = [0] * 26
        for c in s:
            count[ord(c) - ord("a")] += 1

        res = []
        i = 0

        for c in range(26):
            # pop all the chars from stack that are less than or equal to current char
            while stack and stack[-1] <= c:
                res.append(chr(stack.pop() + ord("a")))

            # move forward and print all met current chars
            # all other chars add to stack
            while count[c] > 0:
                count[ord(s[i]) - ord("a")] -= 1

                if ord(s[i]) - ord("a") != c:
                    stack.append(ord(s[i]) - ord("a"))
                else:
                    res.append(s[i])

                i += 1

        while stack:
            res.append(chr(stack.pop() + ord("a")))

        return "".join(res)

